compiler/pipes/calc-bad-vars.cpp

```diff
@@ -118,22 +119,66 @@ struct FuncCallGraph {
 class CalcBadVars {
 private:
   class MergeBadVarsCallback : public MergeReachalbeCallback<FunctionPtr> {
-    IdMap<vector<VarPtr>> tmp_vars;
+    IdMap<vector<VarPtr>> modified_global_vars;
   public:
-    explicit MergeBadVarsCallback(IdMap<vector<VarPtr>> tmp_vars) : tmp_vars(std::move(tmp_vars)) {}
+    explicit MergeBadVarsCallback(IdMap<vector<VarPtr>> &&modified_global_vars) :
+      modified_global_vars(std::move(modified_global_vars)) {}
+
+    // here we calculate "bad vars" for a function — they will be used in check-ub.cpp
+    // "bad vars" are modified non-primitive globals, that can lead to a potential ub when modified and accessed in subcalls
+    // this method just unions all edges' bad_vars to the current function
+    // this works in a single thread and uses some heuristic for speedup, but generally it means
+    // "for f in component { for e in edge { f.bad_vars += e.bad_vars } }"
     void for_component(const vector<FunctionPtr> &component, const vector<FunctionPtr> &edges) override {
-      std::unordered_set<VarPtr> bad_vars_uniq;
+      // optimization 1: lots of functions don't modify globals at all
+      // then we leave f->bad_vars nullptr
+      bool empty = vk::all_of(component, [&](FunctionPtr f) { return modified_global_vars[f].empty(); })
+                && vk::all_of(edges, [&](FunctionPtr e) { return e->bad_vars == nullptr; });
+      if (empty) {
+        return;
+      }
 
-      for (FunctionPtr f : component) {
-        bad_vars_uniq.insert(tmp_vars[f].begin(), tmp_vars[f].end());
+      // optimization 2: maybe, one of the edges has bad_vars that includes all other edges
+      // then we just assign f->bad_vars = largest_edge->bad_vars
+      std::unordered_set<VarPtr> *largest_e{nullptr};
+      for (FunctionPtr e : edges) {
+        if (e->bad_vars != nullptr && (!largest_e || e->bad_vars->size() > largest_e->size())) {
+          largest_e = e->bad_vars;
+        }
       }
+      if (largest_e) {
+        bool all_in_largest_e = vk::all_of(component, [&](FunctionPtr f) {
+          return vk::all_of(modified_global_vars[f], [&](VarPtr v_in_self) {
+            return largest_e->find(v_in_self) != largest_e->end();
+          });
+        });
+        all_in_largest_e &= vk::all_of(edges, [&](FunctionPtr e) {
```

Comment:
I'm not 100% against joining bool results with `&`, but when it leads to another edges traversal I'm more inclined that it's not the most beautiful thing.

It's up to you whether it should be changed or not.

Reply:
How'd you suggest to correct this not very beautiful code? ) I tried to use
```
bool all_in_largest_e = vk::all_of(
  ...
) && vk::all_of( 
  ...
);
```
but it seems even less beautiful than the final way.